src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs

```diff
@@ -696,7 +696,26 @@ private BoundBlock RewriteBlock(BoundBlock node, ArrayBuilder<BoundExpression> p
 
             if (prologue.Count > 0)
             {
-                newStatements.Add(new BoundSequencePoint(null, null) { WasCompilerGenerated = true });
+                // Add hidden sequence point, prologue -- closure initialization -- doesn't map to source.
+
+                // However, when the closure is used inside of conditional blocks like 'if (b) {...}',
+                // we need to put a regular seqeunce point to closure initialization 
+                // associated with the closure scope start ('{'), so IP can move to '{' without skipping initialization.
+
+                var initialization = prologue.First();
+                var block = initialization.Syntax as BlockSyntax;
+                if (block != null && (block.Parent.IsKind(SyntaxKind.IfStatement) || block.Parent.IsKind(SyntaxKind.ElseClause)))
```

Comment:
+1

The sequence points needs to be associated with the closure scope syntax span start in general, not just in one particular case.